include/pybind11/pytypes.h

```diff
@@ -334,17 +334,19 @@ class object : public handle {
     }
 
     object &operator=(const object &other) {
-        other.inc_ref();
-        // Use temporary variable to ensure `*this` remains valid while
-        // `Py_XDECREF` executes, in case `*this` is accessible from Python.
-        handle temp(m_ptr);
-        m_ptr = other.m_ptr;
-        temp.dec_ref();
+        if (!this->is(other)) {
```

Comment:
Is this purely an optimization that came to your attention because of the current focus on in-place operators, or is this `if` critical in some way?
A terse comment might be useful for readers, e.g.
```
if (!this->is(other)) { // Optimization for in-place operators.
```
if that's true, or something else if not.

Reply:
It came to my attention due to inplace-assigment, but it's a more general optimization.